src/exchange/bitfinex/mod.rs

```diff
@@ -0,0 +1,624 @@
+                            continue;
+                        }
+                    },
+                    Some(Ok(WsMessage::Close(close_frame))) => {
+                        break Err(SocketError::Subscribe(format!(
+                            "received WebSocket CloseFrame: {:?}", close_frame
+                        )))
+                    },
+                    _ => continue,
+                }
+            }
+        }
+    }
+}
+
+impl Bitfinex {
+    /// [`Bitfinex`] trades channel name.
+    ///
+    /// See docs: <https://docs.bitfinex.com/reference/ws-public-trades>
+    const CHANNEL_TRADES: &'static str = "trades";
+
+    /// [`Bitfinex`] candles channel name.
+    ///
+    /// See docs: <https://docs.bitfinex.com/reference/ws-public-candles>
+    const CHANNEL_CANDLES: &'static str = "candles";
+
+    /// Build the initial [`Bitfinex`] compatible [`SubscriptionId`] using the channel & market
+    /// provided. Note this is only used to confirm subscription responses. The [`SubscriptionId`]
+    /// is then replaced with the numeric `CHANNEL_ID` [`Bitfinex`] uses as message identifiers
+    /// (not persistent across connection).
+    ///
+    /// This `SubscriptionId(CHANNEL_ID)` is then used to
+    /// associated [`Bitfinex`] data structures received over the Websocket with it's original
+    /// Barter [`Subscription`].
+    ///
+    /// eg/ SubscriptionId("trades|tBTCUSD")
+    fn subscription_id(channel: &str, market: &str) -> SubscriptionId {
+        SubscriptionId::from(format!("{channel}|{market}"))
+    }
+
+    /// Determine the [`Bitfinex`] channel metadata associated with an input Barter
+    /// [`Subscription`]. This includes the `Bitfinex` &str channel, and a `String` market
+    /// identifier. Both are used to build a `Bitfinex` subscription payload.
+    ///
+    /// Example Ok Return: Ok("trades", "tBTCUSD")
+    /// Example Ok Return: Ok("candles", "trade:1m:tBTCUSD")
+    /// where channel == "trades" & market == "tBTCUSD".
+    fn build_channel_meta(sub: &Subscription) -> Result<(&str, String), SocketError> {
+        let sub = sub.validate()?;
+
+        // Determine Bitfinex channel
+        let channel = match &sub.kind {
+            SubKind::Trade => Self::CHANNEL_TRADES,
+            SubKind::Candle(_) => Self::CHANNEL_CANDLES,
+            other => {
+                return Err(SocketError::Unsupported {
+                    entity: Self::EXCHANGE.as_str(),
+                    item: other.to_string(),
+                })
+            }
+        };
+
+        // Determine Bitfinex market using the Subscription Instrument
+        let market = match &sub.instrument.kind {
+            InstrumentKind::Spot => format!(
+                "t{}{}",
+                sub.instrument.base.to_string().to_uppercase(),
+                sub.instrument.quote.to_string().to_uppercase()
+            ),
+            InstrumentKind::FuturePerpetual => todo!(),
```

Comment:
Perhaps this should be returning a SocketError if you are not planning to implement in this MR?

ie/
```suggestion
            other => {
                return Err(SocketError::Unsupported {
                    entity: Self::EXCHANGE.as_str(),
                    item: other.to_string(),
                })
            }
```